File: crawlernest/crawlernest-web/src/lib/datasetScope.ts
/**
 * What the warehouse holds, as the frontend sees it.
 *
 * Mirrors `DATASET_YEARS` and `DEFAULT_RANKING_YEAR` in
 * `crawlernest/core/dataset.py` and `DatasetScope.DATASET_YEARS` in Java.
 * Python's `test_caveat_contract` compares the three, so a change here needs the
 * same change there. Changing it is a data-migration step, not an edit.
 */

/**
 * Which editions each source actually covers, newest first.
 *
 * Mirrors `DATASET_COVERAGE` in `crawlernest/core/dataset.py` and
 * `DatasetScope.DATASET_COVERAGE` in Java. ARWU reaches back to 2015; QS and THE
 * begin at 2025. The snapshot caveat names a source, so rendering it over every
 * held edition would tell a reader that ten editions with no QS row in them are
 * a snapshot of the QS published tables.
 */
export const DATASET_COVERAGE: Readonly<Record<string, readonly number[]>> = {
  QS: [2026, 2025],
  THE: [2026, 2025],
  ARWU: [2026, 2025, 2024, 2023, 2022, 2021, 2020, 2019, 2018, 2017, 2016, 2015],
};

/** Ranking sources ingested, widest first. Mirrors `DATASET_SOURCES` in Python. */
export const DATASET_SOURCES: readonly string[] = ["QS", "THE", "ARWU"];

/**
 * Every ranking edition the warehouse holds, newest first.
 *
 * The union of `DATASET_COVERAGE`, written out because `test_caveat_contract`
 * reads this literal and compares it with the Python and Java ones.
 */
export const DATASET_YEARS: readonly number[] = [
  2026, 2025, 2024, 2023, 2022, 2021, 2020, 2019, 2018, 2017, 2016, 2015,
];

/**
 * The sources holding ranks for an edition, in `DATASET_SOURCES` order. Empty
 * for an edition the warehouse does not hold.
 */
export function sourcesForYear(year: number): string[] {
  return DATASET_SOURCES.filter((source) => (DATASET_COVERAGE[source] ?? []).includes(year));
}

/** The year a query uses when the caller names none: the newest edition held. */
export const DEFAULT_RANKING_YEAR: number = Math.max(...DATASET_YEARS);

/** The query parameter that carries the selected edition, so a shared link keeps it. */
export const YEAR_QUERY_PARAM = "year";

export type SelectedYear = {
  /** The edition to read: always one the warehouse holds. */
  year: number;
  /** What the URL asked for, verbatim; null when it named none. */
  requested: string | null;
  /** False when the URL named an edition that is not held, so `year` is the default instead. */
  requestedHeld: boolean;
};

/**
 * The edition a page should read for a `?year=` value. Same rule as
 * `DatasetScope.resolveRankingYear` in Java and `resolve_ranking_year` in Python,
 * except that where the API reads nothing for an unheld year, a page falls back to
 * the default and says so -- an empty table with no explanation reads as "no
 * universities", not as "no such edition".
 */
export function resolveSelectedYear(raw: string | null | undefined): SelectedYear {
  const requested = raw == null || raw.trim() === "" ? null : raw.trim();
  if (requested === null) {
    return { year: DEFAULT_RANKING_YEAR, requested: null, requestedHeld: true };
  }
  const parsed = /^\d{4}$/.test(requested) ? Number(requested) : NaN;
  if (DATASET_YEARS.includes(parsed)) {
    return { year: parsed, requested, requestedHeld: true };
  }
  return { year: DEFAULT_RANKING_YEAR, requested, requestedHeld: false };
}

/** `href` with `?year=` set, keeping any query it already has. */
export function hrefWithYear(href: string, year: number): string {
  const [path, query = ""] = href.split("?", 2);
  const params = new URLSearchParams(query);
  params.set(YEAR_QUERY_PARAM, String(year));
  return `${path}?${params.toString()}`;
}

/**
 * An in-app link that keeps the edition being viewed. The default edition needs
 * no parameter -- a plain link already shows it -- so only another edition is
 * written into the URL. One rule for every link between year-aware pages.
 */
export function hrefForEdition(href: string, year: number): string {
  return year === DEFAULT_RANKING_YEAR ? href : hrefWithYear(href, year);
}

File: crawlernest/servise_for_java/src/main/java/clawer/service/DatasetScope.java
package clawer.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.OptionalInt;
import java.util.stream.Collectors;

/**
 * Which ranking editions a read is allowed to see.
 *
 * <p>Every multi-year relation -- {@code analytics.aggregated_rankings},
 * {@code v_aggregated_rankings_latest}, {@code v_recommendation_candidates_latest},
 * {@code warehouse.ranking_record}, the model predictions -- holds one row per
 * university <em>per edition</em>. A read that does not name an edition reads all
 * of them, and three things go wrong the moment a second edition is loaded: each
 * university appears once per edition, "the newest row for this university"
 * silently becomes an older edition's rank for any university the current edition
 * lacks, and an edition loaded but not yet released (a shadow ingest) is served as
 * if it were live.
 *
 * <p>So no serving read decides its own edition. It asks this:
 * <ul>
 *   <li>no year requested: the default edition, the newest one held;</li>
 *   <li>a held year requested: that year;</li>
 *   <li>any other year: nothing. The caller returns an empty result rather than
 *       querying, so a shadow edition is invisible even to {@code ?year=2025}.
 *       There is no exception: nothing maps one to a 4xx here, and an empty result
 *       is what a year with no rows has always returned.</li>
 * </ul>
 *
 * <p>{@link #DATASET_YEARS} is the release's list and mirrors
 * {@code crawlernest/core/dataset.py}; {@code test_caveat_contract.py} compares
 * the literals. The {@code crawlernest.dataset.years} property exists for
 * integration tests, whose fixtures live in an edition no crawl writes (2099) so
 * that the live warehouse they run against cannot leak into their assertions.
 * Setting it in production would contradict the caveats, which render from the
 * static list.
 */
@Component
public class DatasetScope {

    /**
     * Which editions each source actually covers, newest first.
     *
     * <p>Mirrors {@code DATASET_COVERAGE} in {@code crawlernest/core/dataset.py}.
     * ARWU reaches back to 2015; QS and THE begin at 2025. A flat list of years
     * cannot say that, and the snapshot caveat names a source -- rendering it over
     * every held edition would claim QS published tables for ten editions holding
     * no QS row.
     */
    public static final Map<String, List<Integer>> DATASET_COVERAGE = Map.of(
            "QS", List.of(2026, 2025),
            "THE", List.of(2026, 2025),
            "ARWU", List.of(2026, 2025, 2024, 2023, 2022, 2021, 2020, 2019, 2018, 2017, 2016, 2015));

    /** Ranking sources ingested, widest first. Mirrors {@code DATASET_SOURCES} in Python. */
    public static final List<String> DATASET_SOURCES = List.of("QS", "THE", "ARWU");

    /**
     * Every ranking edition the warehouse holds, newest first. Changing it is a data migration.
     *
     * <p>The union of {@link #DATASET_COVERAGE}, written out because
     * {@code test_caveat_contract.py} reads this literal and compares it with the
     * Python and TypeScript ones; {@link #datasetYearsMatchCoverage} keeps the
     * literal honest against the map beside it.
     */
    public static final List<Integer> DATASET_YEARS = List.of(
            2026, 2025, 2024, 2023, 2022, 2021, 2020, 2019, 2018, 2017, 2016, 2015);

    static {
        if (!datasetYearsMatchCoverage()) {
            throw new IllegalStateException(
                    "DATASET_YEARS is not the union of DATASET_COVERAGE; one of the two is wrong");
        }
    }

    /** True when the literal list is exactly the editions some source covers. */
    static boolean datasetYearsMatchCoverage() {
        List<Integer> union = DATASET_COVERAGE.values().stream()
                .flatMap(List::stream)
                .distinct()
                .sorted(Comparator.reverseOrder())
                .toList();
        return union.equals(DATASET_YEARS);
    }

    /**
     * The sources holding ranks for an edition, in {@link #DATASET_SOURCES} order.
     * Empty for an edition the warehouse does not hold. {@code sources_for_year} in
     * Python and {@code sourcesForYear} in TypeScript are the same rule.
     */
    public static List<String> sourcesFor(int year) {
        return DATASET_SOURCES.stream()
                .filter(source -> DATASET_COVERAGE.getOrDefault(source, List.of()).contains(year))
                .toList();
    }

    /** The edition a read uses when the caller names none. */
    public static final int DEFAULT_RANKING_YEAR = Collections.max(DATASET_YEARS);

    private final List<Integer> heldYears;
    private final int defaultRankingYear;

    /** The release's editions. */
    public DatasetScope() {
        this("");
    }

    @Autowired
    public DatasetScope(@Value("${crawlernest.dataset.years:}") String override) {
        List<Integer> years = parse(override);
        this.heldYears = years.isEmpty() ? DATASET_YEARS : years;
        this.defaultRankingYear = Collections.max(this.heldYears);
    }

    /** The scope a test or a non-Spring caller gets without configuration. */
    public static DatasetScope standard() {
        return new DatasetScope();
    }

    public List<Integer> heldYears() {
        return heldYears;
    }

    public int defaultRankingYear() {
        return defaultRankingYear;
    }

    public boolean isHeld(int year) {
        return heldYears.contains(year);
    }

    /** The edition to read for a request, or empty when the requested edition is not held. */
    public OptionalInt resolveRankingYear(Integer requestedYear) {
        if (requestedYear == null) {
            return OptionalInt.of(defaultRankingYear);
        }
        return isHeld(requestedYear) ? OptionalInt.of(requestedYear) : OptionalInt.empty();
    }

    /**
     * The held editions as a PostgreSQL array literal, bound as {@code ?::int[]}.
     * Built from integers, so it cannot carry anything but digits and commas.
     */
    public String heldYearsSqlArray() {
        return heldYears.stream().map(String::valueOf).collect(Collectors.joining(",", "{", "}"));
    }

    private static List<Integer> parse(String override) {
        if (override == null || override.isBlank()) {
            return List.of();
        }
        return Arrays.stream(override.split(","))
                .map(String::trim)
                .filter(value -> !value.isEmpty())
                .map(Integer::valueOf)
                .distinct()
                .sorted(Comparator.reverseOrder())
                .toList();
    }
}
